runtime/gc_vlhgc/MemorySubSpaceTarok.cpp

```diff
@@ -1355,18 +1384,169 @@ MM_MemorySubSpaceTarok::calculateExpandSize(MM_EnvironmentBase *env, UDATA bytes
+		Trc_MM_MemorySubSpaceTarok_getHeapSizeWithinBounds_2(env->getLanguageVMThread(), heapSizePercentChange);
+		recommendedHeapSize = (uintptr_t)(heapSizePercentChange * recommendedHeapSize);
+	}
+
+	return recommendedHeapSize;
+}
+
+
+double 
+MM_MemorySubSpaceTarok::calculateGcPctForHeapChange(MM_EnvironmentBase *env, intptr_t heapSizeChange)
+{
+	MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+
+	/* 
+	 * Since we only attempt to resize after PGC's and Global GC's, if the PGC count has increased, it means we are resizing after a PGC
+	 * and the GMP and PGC statistics we have collected, are useable. Otherwise, use backup methods which are independant of heapSizeChange
+	 */
+
+	if (_previouslyObservedPGCCount != envVLHGC->_heapSizingData.pgcCountSinceGMPEnd) {
+		/* Resizing after a PGC */
+
+		uintptr_t pgcCount = envVLHGC->getRepresentativePgcPerGmpCount();
+
+		if ((0 == pgcCount) && (0 == _lastObservedGcPercentage)) {
+			/* Very first time we are resizing, assume GC % is heapExpansionGCTimeThreshold */
+			_lastObservedGcPercentage = (double)_extensions->heapExpansionGCTimeThreshold;
+
+		} else {
+
+			if (0 != heapSizeChange) {
+				/* 
+				 * Determine what the gc cpu % would be if we changed the heap by heapSizeChange bytes
+				 * Main idea is that the change in number of pgc's (per gmp) will be proportional to how much free tenure changes;
+				 */
+				MM_EnvironmentVLHGC *envVLHGC = (MM_EnvironmentVLHGC *)env;
+				uintptr_t currentFreeTenure = (uintptr_t)envVLHGC->_heapSizingData.freeTenure;
+				uintptr_t potentialFreeTenure = currentFreeTenure + heapSizeChange;
+				if (heapSizeChange <= (-1 * (intptr_t)currentFreeTenure) ) {
+					/* If we try to shrink too much, too fast, tenure will be way too small, causing lots of GC work */
+					potentialFreeTenure = 1;
+				}
+				pgcCount = (uintptr_t)(((double)potentialFreeTenure / currentFreeTenure) * pgcCount);
+			}
+
+			double gcActiveTime = (double)(pgcCount * envVLHGC->_heapSizingData.avgPgcTimeUs) + envVLHGC->_heapSizingData.gmpTime; 
+			double gcInterval = (double)(pgcCount  * (envVLHGC->_heapSizingData.avgPgcTimeUs + envVLHGC->_heapSizingData.avgPgcIntervalUs)) + envVLHGC->_heapSizingData.gmpTime;
+			double gcActiveRatio = gcActiveTime / gcInterval;
+
+			_lastObservedGcPercentage = gcActiveRatio * 100;
+		}
 
-UDATA
+	} else {
+		/* 
+		 * Resizing the heap after global GC. The PGC and GMP timing data MAY no longer be accurate - look at backup methods. 
+		 * The values below do not change if heapSizeChange != 0, since global GC is proportional to live data in heap, rather than heap size. 
+		 * This means changing the heap size will not change the gc% by much
+		 */
+		if(NULL != _collector) {		
+			_lastObservedGcPercentage = (double)_collector->getGCTimePercentage(env);	
+		} else {	
+			_lastObservedGcPercentage= (double)_extensions->getGlobalCollector()->getGCTimePercentage(env);	
```

Comment:
```suggestion
			_lastObservedGcPercentage = (double)_extensions->getGlobalCollector()->getGCTimePercentage(env);	
```

Reply:
in which case we wouldn't have _collector?

Reply:
I'm not sure, however, the previous way to get GC percentage had this same condition (checking if there is a `_collector`), so I'm guessing whoever originally wrote that piece of code did so for a reason.

See `MM_MemorySubSpaceTarok::checkForRatioExpand()` in the changed files to see this usage.